### PR TITLE
fix(41268): Replace leading tabs in diagnostic code with spaces

### LINKED ISSUE #41268: Not all leading tab characters in diagnostic messages are replaced with spaces
The error diagnostics that are used by default do not replace ALL leading tab characters with a single space.

As one would expect from seeing this code line in souce code:
https://github.com/microsoft/TypeScript/blob/85ef91e2233fa384e980005c109fadc04376ce3e/src/compiler/program.ts#L410

This line only replaces the first leading tab character. Which causes way too much leading whitespace printed in the logs when errors occur during tsc build.

**TypeScript Version:**  4.0.3

**Search Terms:**
tabs diagnostics, reporter, formatCodeSpan, formatDiagnosticsWithColorAndContext, leading tabs,

**Expected behavior:**
Source code has four leading tabs for indentation.

`[line number] [gutter seperator] [space] [space] [space] [space] [line content]`

image

**Actual behavior:**
`[line number] [gutter seperator] [space] [tab] [tab] [tab] [line content]`

image

**PR**
I can submit a PR for this that would change this line:
`lineContent = lineContent.replace("\t", " ");    // convert tabs to single spaces`
to:
`lineContent = lineContent.replace(/\t/gy, " "); // convert all leading tabs to single spaces`

### PR DESCRIPTION
**Pull request by thdk**
When tabs are used for indentation in source files,
they should be replaced by spaces when printed in diagnostic messages.

Current code only replaced the first leading tab character.
This commit will now replace all leading tabs when printing source code
lines in diagnostic messages.

* [x] There is an associated issue in the `Backlog` milestone (**required**)
* [x] Code is up-to-date with the `master` branch
* [x] You've successfully run `gulp runtests` locally
* [ ] There are new or updated unit tests validating the change

If tests are needed, please guide me to some example test for diagnostics.

Fixes #41268

### CODE REVIEW

File: src/compiler/program.ts [hunk 1]
```diff
@@ -406,8 +406,12 @@ namespace ts {
...
             const lineEnd = i < lastLineInFile ? getPositionOfLineAndCharacter(file, i + 1, 0) : file.text.length;
             let lineContent = file.text.slice(lineStart, lineEnd);
-            lineContent = lineContent.replace(/\s+$/g, "");  // trim from end
-            lineContent = lineContent.replace("\t", " ");    // convert tabs to single spaces
+
+            // Trim from end
+            lineContent = lineContent.replace(/\s+$/g, "");
+
+            // Convert all leading tabs to single spaces
+            lineContent = lineContent.replace(/\t/gy, " ");
```
Comment (RyanCavanaugh) [inline src/compiler/program.ts:414]: Why the `y` flag?

Comment (thdk) [reply src/compiler/program.ts:414]: I though that only leading tabs in source code (used for indentation) should be replaced by spaces. I was not sure if replacing all tabs would be ok.

```
// example source code with leading tabs and tabs used elsewhere in the line

			private readonly foo: string;			// replace only leading tabs
```
`/\t/gy` prints:
```
   private readonly foo: string;			// replace only leading tabs
```

`/\t/g` prints:
```
   private readonly foo: string;   // replace only leading tabs
```
If you are ok to replace all tabs with spaces I can of course remove the `y` flag.

Comment (sandersn) [reply src/compiler/program.ts:414]: So the iterative nature of `replace` combined with `y=sticky` means that the regex stops matching after the initial run, right? If so, that's pretty subtle.

I'd rather have this explicitly documented, adding "(the sticky flag 'y' means that the regex stops matching after the initial run of tabs)" or something like it.

Also, [the MDN page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/sticky) says the `g` is ignored for regexes with `y`.

**Review by sandersn — CHANGES_REQUESTED** [review]

File: src/compiler/program.ts (hunk 1 above)
Comment (thdk) [reply src/compiler/program.ts:414]: To be honest, I'm not so confident with regular expressions that it took me some time to come up with this solution to replace **all leading** tab characters. I was surprised not to find an example where this is done. I've seen the note on MDN as well but without the _global_ flag, only the first tab character was replaced in my test.

```javascript
// input text with two leading tab characters and two tab characters elsewhere in the text
var input="     Some text       containing tabs";

// A:
console.log(input.replace(/\t/g, "LEADING_TAB_"));  
// prints: "LEADING_TAB_LEADING_TAB_Some textLEADING_TAB_LEADING_TAB_containing tabs"

// B:
console.log(input.replace(/\t/, "LEADING_TAB_"));  
// prints: "LEADING_TAB_   Some text       containing tabs"

// C:
console.log(input.replace(/^(\t)*/, "LEADING_TAB_"));  
// prints: "LEADING_TAB_Some text      containing tabs"

// D:
console.log(input.replace(/\t/gy, "LEADING_TAB_"));  
// prints: "LEADING_TAB_LEADING_TAB_Some text      containing tabs"

// E:
console.log(input.replace(/\t/y, "LEADING_TAB_"));  
// prints: "LEADING_TAB_   Some text       containing tabs"
```
I've created a [jsfiddle](https://jsfiddle.net/onebcvu4/2/) and asked about **why this works** on [stackoverflow](https://stackoverflow.com/questions/64982135/regex-to-replace-all-leading-tab-characters-each-one-with-a-single-space). 

While writing this comment, my question was [answered](https://stackoverflow.com/a/64982184/681803).

Now I just need to find a good comment that clarifies the use of both _sticky_ and _global_ for this line.

Does this do it for you?

```javascript
// Convert all leading tabs to single spaces
// using both g and y flag to repeatedly match a \t at lastIndex (default: 0) and replace it until it runs out of \t at lastIndex
lineContent = lineContent.replace(/\t/gy, " "); 
```

Comment (thdk) [reply src/compiler/program.ts:414]: Not sure though about what's the max number of characters per line at Typescript and guideline for multi line comments.

**Comment by thdk** [comment]
Are there still changes expected from me for this issue?

**Review by sandersn — APPROVED** [review]

**Comment by sandersn** [comment]
Looks like github's merge UI (on windows?) changed line endings or something. I'm going to check this out locally on a machine I trust to see if I can fix that before merging.

**Review by sandersn — CHANGES_REQUESTED** [review]
Needs a test.

**Comment by sandersn** [comment]
@sheetalkamat good point -- what kind of test is appropriate here?

**Comment by sheetalkamat** [comment]
https://github.com/microsoft/TypeScript/tree/master/src/testRunner/unittests/tsc is where we run tests like tsc and should show the issue and fix

**Comment by sandersn** [comment]
It's been a year, so I'm going to close this to reduce the number of old open PRs. @thdk Let me know if you want to keep working on this.

**Pull request closed without merging**